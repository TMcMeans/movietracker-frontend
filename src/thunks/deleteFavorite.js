import { hasErrored } from '../actions';

export const deleteFavorite = (userId, movieId) => {
  return async (dispatch) => {
    try {
      const response = await fetch(`http://localhost:3000/api/users/${userId}/favorites/${movieId}`, {
        method: 'DELETE',
        headers: {
          'Content-Type': 'application/json'
        }
      });
      if (!response.ok) {
        throw Error(response.statusText);
      }
      console.log(response)
    } catch(err) {
      dispatch(hasErrored(err.message));
    }
  }
}