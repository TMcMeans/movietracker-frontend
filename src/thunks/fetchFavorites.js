import { hasErrored, retrieveAllFavorites } from "../actions";

export const fetchFavorites = (userId) => {
  return async(dispatch) => {
    try {
      const response = await fetch(`http://localhost:3000/api/users/${userId}/favorites`);
      if (!response.ok) {
        throw Error(response.statusText)
      }
      const result = await response.json()
      console.log(result)
      dispatch(retrieveAllFavorites(result.data))
    } catch(err) {
      dispatch(hasErrored(err.message))
    }
  }
}