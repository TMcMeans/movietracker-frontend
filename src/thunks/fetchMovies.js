import { addMovies, hasErrored } from '../actions';

export const fetchMovies = url => {
  return async dispatch => {
    try {
      const response = await fetch(url);
      if (!response.ok) {
        console.log(response)
        throw Error(response.statusText);
      }
      const movies = await response.json();
      dispatch(addMovies(movies.results));
    } catch (error) {
      dispatch(hasErrored(error.message));
    }
  };
};