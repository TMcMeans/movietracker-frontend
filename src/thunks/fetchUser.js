import { hasErrored, loginUser, addMessage } from "../actions";

export const fetchUser = (email, password) => {
  return async (dispatch) => {
    try {
      const response = await fetch('http://localhost:3000/api/users', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          email,
          password
        })
      });
      if(!response.ok) {
        dispatch(addMessage('Password or Email does not Match'))
        throw Error(response.statusText)
      }
      const result = await response.json()
      dispatch(loginUser({name: result.data.name, id: result.data.id}))
      dispatch(addMessage('Sucess! You are now Logged in'))
    } catch(err) {
      console.log(err)
      dispatch(hasErrored(err.message))
      dispatch(addMessage('Internal Server Error, Failed to Login'))
    }
  }
}