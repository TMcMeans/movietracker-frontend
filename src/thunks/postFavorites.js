import { hasErrored, addMessage } from '../actions';

export const postFavorites = (movie, userId) => {
  return async dispatch => {
    try {
      const response = await fetch(
        'http://localhost:3000/api/users/favorites/new',
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json'
          },
          body: JSON.stringify({
            ...movie,
            user_id: userId
          })
        }
      );
      if (!response.ok) {
        throw Error(response.statusText);
      }
      const result = await response.json();
      console.log(result)
      dispatch(addMessage('Movie is now being Stalked'));
    } catch (err) {
      dispatch(hasErrored(err.message));
      dispatch(addMessage('Internal Server Error, Failed to Stalk Movie'))
    }
  };
};
