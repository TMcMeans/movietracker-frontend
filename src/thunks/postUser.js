import { hasErrored, loginUser } from "../actions";

export const postUser = (email, password, name) => {
  return async(dispatch) => {
    try {
      const response = await fetch('http://localhost:3000/api/users/new', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          name,
          email,
          password
        })
      });
      if(!response.ok) {
        throw Error(response.statusText)
      }
      const result = await response.json()
      dispatch(loginUser({name, id: result.id}))
    } catch(err) {
      dispatch(hasErrored(err.message))
    }
  }
}