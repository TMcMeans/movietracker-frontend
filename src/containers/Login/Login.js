import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router';
import { fetchUser } from '../../thunks/fetchUser'
import { postUser } from '../../thunks/postUser'

export class Login extends Component {
  constructor() {
    super();
    this.state = {
      name: '',
      email: '',
      password: '',
      errorMessage: '',
      newUser: false
    }
  }

  handleInputChange = e => {
    const { name, value } = e.target;
    this.setState({
      [name]: value
    });
  }
  
  handleSubmit = async e => {
    e.preventDefault();
    const { email, password, newUser, name } = this.state;
    if (newUser) {
      this.props.addNewUserToStore(email, password, name)
    } else {
     this.props.addUserToStore(email, password)
    }
  }

  handleNewUser = (e) => {
    e.preventDefault();
    this.setState({
      newUser: true
    })
  }

  render() {
    const { email, password, name, newUser } = this.state;
    let nameInput;
    let buttonText;
    let page;
    let createNewUser = <p className="create-account"onClick={this.handleNewUser}>Click here to create an account</p>
    if (newUser) {
      buttonText = 'Create New Account'
      nameInput = (
        <input
              type="text"
              placeholder="Enter Name"
              value={name}
              onChange={this.handleInputChange}
              name="name"
            />
            )
      createNewUser = ''
    } else {
      buttonText = 'Submit';
      nameInput = '';
    }
    if(this.props.user) {
      page = <Redirect to='/' />
    }

    return (
      <div className="form-container">
        <form onSubmit={this.handleSubmit}>
          <h1>LOGIN</h1>
          {nameInput}
          <input
            type="email"
            placeholder="E-mail"
            value={email}
            onChange={this.handleInputChange}
            name="email"
          />
          <input
            type="password"
            placeholder="password"
            value={password}
            onChange={this.handleInputChange}
            name="password"
          />
          <button>{buttonText}</button>
        </form>
        {createNewUser}
        <p>{this.state.errorMessage}</p>
        {page}
      </div>
    );
  }
}
export const mapStateToProps = state => ({
  user: state.user
})

export const mapDispatchToProps = dispatch => ({
  addUserToStore: (email, password) => dispatch(fetchUser(email, password)),
  addNewUserToStore: (email, password, name) => dispatch(postUser(email, password, name))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Login);
