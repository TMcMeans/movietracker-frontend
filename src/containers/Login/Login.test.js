import { shallow } from 'enzyme'
import React from 'react';
import { Login, mapStateToProps, mapDispatchToProps } from './Login';
import { fetchUser } from '../../thunks/fetchUser';
import { postUser } from '../../thunks/postUser';

jest.mock('../../thunks/fetchUser');
jest.mock('../../thunks/postUser');

describe('Login container', () => {
  it('should match the snapshot', () => {
    let wrapper = shallow(<Login />, {disableLifecycleMethods: true})
    expect(wrapper).toMatchSnapshot()
  })
})

describe('handleInputChange', () => {
  it('should update state on change', () => {
    let wrapper = shallow(<Login />, {disableLifecycleMethods: true});
    wrapper.find('#email-input').simulate('change', 
    { 
      target: 
      { 
        value: 'john@gmail.com',
        name: 'email' 
      } 
    })  
    wrapper.find('#password-input').simulate('change', 
    { 
      target: 
      { 
        value: 'Hello',
        name: 'password' 
      } 
    })   
    const expected = {
      name: '',
      email: 'john@gmail.com',
      password: 'Hello',
      errorMessage:'',
      newUser: false
    }
    expect(wrapper.state()).toEqual(expected)
  })
})

describe('handleSubmit', () => {
  let wrapper;
  it('should call handleSubmit and should call addUserToStore if !newUser', () => {
    const mockAddUserToStore = jest.fn()
    wrapper = shallow(<Login addUserToStore={mockAddUserToStore}/>, {disableLifecycleMethods: true});
    wrapper.handleSubmit = jest.fn()
    wrapper.addUserToStore = mockAddUserToStore

    wrapper.find('form').simulate('submit', {
      preventDefault: () => wrapper.handleSubmit()})

    expect(wrapper.handleSubmit).toHaveBeenCalled()
    expect(wrapper.addUserToStore).toHaveBeenCalled();
  });

  it('should call addNewUserToStore if new user',() => {
    const mockAddUserToStore = jest.fn()
    wrapper = shallow(<Login addNewUserToStore={mockAddUserToStore}/>, {disableLifecycleMethods: true});
    wrapper.handleSubmit = jest.fn()
    wrapper.addNewUserToStore = mockAddUserToStore
    wrapper.setState({newUser: true})

    wrapper.find('form').simulate('submit', {
      preventDefault: () => wrapper.handleSubmit()})
    expect(wrapper.addNewUserToStore).toHaveBeenCalled()
  })
})

describe('handleNewUser', () => {
  it('should toggle the state of newUser', () => {
    let wrapper = shallow(<Login />, {disableLifecycleMethods: true});
    
    wrapper.find('.create-account').simulate('click', {
      preventDefault: jest.fn()
    })
    expect(wrapper.state().newUser).toEqual(true);

    wrapper.find('.create-account').simulate('click', {
      preventDefault: jest.fn()
    })
    expect(wrapper.state().newUser).toEqual(false);
  })
})

describe('mapStateToProps', () => {
  it('should return an object with user', () => {
    const mockState = {
      movies: [{tile: 'It', id: 0}],
      user: null
    }
    const expected = {
      user: null
    }
    const mappedProps = mapStateToProps(mockState)
    expect(mappedProps).toEqual(expected)
  })
})

describe('mapDispatchToProps', () => {
  const mockDispatch = jest.fn()
  const mappedProps = mapDispatchToProps(mockDispatch);
  const email = 'jack@overlook.com';
  const password = 'password';

  it('calls dispatch with an fetchUser thunks when addUserToStore is called', () => {
    fetchUser.mockImplementation(() => {});
    const expected = fetchUser(email, password);
    mappedProps.addUserToStore('jack@overlook.com', 'password');
    
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });

  it('calls dispatch with an postUser thunks when addNewUserToStore is called', () => {
    postUser.mockImplementation(() => {});
    const userInfo = ('danny@shinning.com', 'password', 'Danny')
    const expected = postUser(userInfo)
    mappedProps.addNewUserToStore('danny@shinning.com', 'password', 'Danny');
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });
});