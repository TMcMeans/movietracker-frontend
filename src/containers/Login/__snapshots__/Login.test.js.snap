// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login container should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <form
          onSubmit={[Function]}
        >
          <h1>
            LOGIN
          </h1>
          
          <input
            id="email-input"
            name="email"
            onChange={[Function]}
            placeholder="E-mail"
            type="email"
            value=""
          />
          <input
            id="password-input"
            name="password"
            onChange={[Function]}
            placeholder="password"
            type="password"
            value=""
          />
          <button>
            Submit
          </button>
        </form>,
        <p
          className="create-account"
          onClick={[Function]}
        >
          Click here to create an account
        </p>,
        <p>
          
        </p>,
        undefined,
      ],
      "className": "form-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1>
              LOGIN
            </h1>,
            "",
            <input
              id="email-input"
              name="email"
              onChange={[Function]}
              placeholder="E-mail"
              type="email"
              value=""
            />,
            <input
              id="password-input"
              name="password"
              onChange={[Function]}
              placeholder="password"
              type="password"
              value=""
            />,
            <button>
              Submit
            </button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "LOGIN",
            },
            "ref": null,
            "rendered": "LOGIN",
            "type": "h1",
          },
          "",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "id": "email-input",
              "name": "email",
              "onChange": [Function],
              "placeholder": "E-mail",
              "type": "email",
              "value": "",
            },
            "ref": Object {
              "current": null,
            },
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "id": "password-input",
              "name": "password",
              "onChange": [Function],
              "placeholder": "password",
              "type": "password",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Submit",
            },
            "ref": null,
            "rendered": "Submit",
            "type": "button",
          },
        ],
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Click here to create an account",
          "className": "create-account",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Click here to create an account",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
        },
        "ref": null,
        "rendered": "",
        "type": "p",
      },
      undefined,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <form
            onSubmit={[Function]}
          >
            <h1>
              LOGIN
            </h1>
            
            <input
              id="email-input"
              name="email"
              onChange={[Function]}
              placeholder="E-mail"
              type="email"
              value=""
            />
            <input
              id="password-input"
              name="password"
              onChange={[Function]}
              placeholder="password"
              type="password"
              value=""
            />
            <button>
              Submit
            </button>
          </form>,
          <p
            className="create-account"
            onClick={[Function]}
          >
            Click here to create an account
          </p>,
          <p>
            
          </p>,
          undefined,
        ],
        "className": "form-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1>
                LOGIN
              </h1>,
              "",
              <input
                id="email-input"
                name="email"
                onChange={[Function]}
                placeholder="E-mail"
                type="email"
                value=""
              />,
              <input
                id="password-input"
                name="password"
                onChange={[Function]}
                placeholder="password"
                type="password"
                value=""
              />,
              <button>
                Submit
              </button>,
            ],
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "LOGIN",
              },
              "ref": null,
              "rendered": "LOGIN",
              "type": "h1",
            },
            "",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "email-input",
                "name": "email",
                "onChange": [Function],
                "placeholder": "E-mail",
                "type": "email",
                "value": "",
              },
              "ref": Object {
                "current": null,
              },
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "password-input",
                "name": "password",
                "onChange": [Function],
                "placeholder": "password",
                "type": "password",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Submit",
              },
              "ref": null,
              "rendered": "Submit",
              "type": "button",
            },
          ],
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Click here to create an account",
            "className": "create-account",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Click here to create an account",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
          },
          "ref": null,
          "rendered": "",
          "type": "p",
        },
        undefined,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
